### PR TITLE
Allow specifying git-host-to-provider-type mapping for on-prem gitlab support

### PR DESCRIPTION
**Pull request by foot**
Adds support for:
- `gitops install` for on-prem gitlab
- oauth for one on-prem host _or_ gitlab.com at a time (globally set via `GITLAB_HOST`)

On oauth with multiple gitlab hosts (requires more work):
  - We must configure a gitlab-app for each gitlab host
  - UI tokens saved in localStorage should be keyed by host, not provider.

On github:
- `gitops install` for on-prem github _might_ work (untested)
- oauth support for on-prem github probably doesn't work?.. (untested)

Closes: #1203 

**What changed?**

- New flag / env var `git-host-types` / `GITOPS_GIT_HOST_TYPES` to specify a mapping
  - `--git-host-types="gl.acme.org=gitlab,gh.acme.org=github"`

**Why?**

Support `gitops install` on-prem non-gitlab.com hosts

**How did you test it?**

- Manually against on-prem gitlab

**Release notes**

Not yet..

**Documentation Changes**

Not yet..

### CODE REVIEW

File: pkg/gitproviders/factory.go [hunk 1]
```diff
@@ -41,28 +41,40 @@ func buildGitProvider(config Config) (gitprovider.Client, string, error) {
...
 		}
-		if config.Hostname != "" {
-			opts = append(opts, gitprovider.WithDomain(config.Hostname))
+
+		// Quirk, see above
+		hostname := gitlab.DefaultDomain
+		if config.Hostname != "" && config.Hostname != gitlab.DefaultDomain {
+			// Quirk, see above
+			hostname = "https://" + config.Hostname
+			opts = append(opts, gitprovider.WithDomain(hostname))
```
Comment (foot) [inline pkg/gitproviders/factory.go:71]: Factor this out?

File: pkg/services/auth/internal/gitlab.go [hunk 2]
```diff
@@ -38,6 +38,10 @@ func getEnvDefault(key, defaultValue string) string {
 	return val
 }
 
+func getGitlabHost() string {
+	return getEnvDefault("GITLAB_HOSTNAME", gitlabHost)
+}
```
Comment (foot) [inline pkg/services/auth/internal/gitlab.go:43]: Allow configuring where to send oauth requests..

File: pkg/gitproviders/repo_url.go [hunk 3]
```diff
@@ -139,18 +150,33 @@ func normalizeRepoURLString(url string, providerName GitProviderName) string {
...
-		trimmed = strings.TrimPrefix(url, sshPrefix)
-	} else if strings.HasPrefix(url, httpsPrefix) {
-		trimmed = strings.TrimPrefix(url, httpsPrefix)
+// ViperGetStringMapString looks up a command line flag or env var in the format "foo=1,bar=2"
+// GetStringMapString tries to JSON decode the env var
+// If that fails (silently), try and decode the classic "foo=1,bar=2" form.
+// https://github.com/spf13/viper/issues/911
+func ViperGetStringMapString(key string) map[string]string {
+	sms := viper.GetStringMapString(key)
+	if len(sms) > 1 {
```
Comment (foot) [inline pkg/gitproviders/repo_url.go:167]:
```suggestion
	if len(sms) > 0 {
```

File: cmd/gitops/root/cmd.go [hunk 4]
```diff
@@ -134,3 +143,12 @@ func configureLogger() {
 		log.SetLevel(log.DebugLevel)
 	}
 }
+
+func initViper(cmd *cobra.Command) error {
+	replacer := strings.NewReplacer("-", "_")
+	viper.SetEnvKeyReplacer(replacer)
+	viper.SetEnvPrefix("GITOPS")
+	viper.AutomaticEnv()
```
Comment (bigkevmcd) [inline cmd/gitops/root/cmd.go:152]: You don't need to do this bit for each command you register?

File: cmd/gitops/root/cmd.go [hunk 5]
```diff
@@ -85,6 +88,11 @@ func RootCmd(client *resty.Client) *cobra.Command {
 `, wego.DefaultNamespace),
 		PersistentPreRun: func(cmd *cobra.Command, args []string) {
 			configureLogger()
+			err := initViper(cmd)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error initing viper: %v\n", err)
+				os.Exit(1)
+			}
```
Comment (bigkevmcd) [inline cmd/gitops/root/cmd.go:95]: You could use [log.Fatalf()](https://pkg.go.dev/log#Fatalf) here, it writes to Stderr, and exits with code 1.

File: pkg/gitproviders/factory.go [hunk 6]
```diff
@@ -41,28 +41,40 @@ func buildGitProvider(config Config) (gitprovider.Client, string, error) {
 		opts := []gitprovider.ClientOption{
 			gitprovider.WithOAuth2Token(config.Token),
 		}
-		if config.Hostname != "" {
-			opts = append(opts, gitprovider.WithDomain(config.Hostname))
+
+		// Quirk of ggp, if using github.com or gitlab.com and you prepend
+		// that with https:// you end up with https://https//github.com !!!
+		// What is that...
+		hostname := github.DefaultDomain
```
Comment (bigkevmcd) [inline pkg/gitproviders/factory.go:48]: How about opening an issue in GGP?

File: pkg/services/auth/internal/gitlab.go (hunk 2 above)
Comment (bigkevmcd) [reply pkg/services/auth/internal/gitlab.go:43]: Should we be documenting this somewhere for users?

File: pkg/gitproviders/repo_url.go [hunk 7]
```diff
@@ -139,18 +150,33 @@ func normalizeRepoURLString(url string, providerName GitProviderName) string {
...
+func ViperGetStringMapString(key string) map[string]string {
+	sms := viper.GetStringMapString(key)
+	if len(sms) > 0 {
+		return sms
 	}
 
-	if trimmed != "" {
-		return fmt.Sprintf("ssh://git@%s.com/%s", providerName, trimmed)
+	ss := viper.GetStringSlice(key)
+	out := make(map[string]string, len(ss))
```
Comment (bigkevmcd) [inline pkg/gitproviders/repo_url.go:172]: I am not sure there's any benefit in providing the `cap` here?

**Review by bigkevmcd — CHANGES_REQUESTED** [review]

File: pkg/gitproviders/repo_url.go (hunk 7 above)
Comment (foot) [reply pkg/gitproviders/repo_url.go:172]: Nope, brought the code over from https://github.com/spf13/pflag/blob/master/string_to_string.go#L90-L98 which might need an attribution actually?

File: pkg/services/auth/internal/gitlab.go (hunk 2 above)
Comment (foot) [reply pkg/services/auth/internal/gitlab.go:43]: Not yet it doesn't feel like a very complete feature. See PR description, it becomes the global HOST for all gitlab oauth requests. I could make some follow up issues though.

File: cmd/gitops/root/cmd.go (hunk 4 above)
Comment (foot) [reply cmd/gitops/root/cmd.go:152]: This is the root command and seems to work for `install` and `ui run` at least so assume it works for others.

File: cmd/gitops/root/cmd.go (hunk 5 above)
Comment (foot) [reply cmd/gitops/root/cmd.go:95]: Nice! I copied the existing pattern in the file to make minimal changes, will test this out too!

**Review by luizbafilho — APPROVED** [review]
Thanks, that looks good.

File: cmd/gitops/root/cmd.go [hunk 8]
```diff
@@ -33,6 +35,18 @@ var options struct {
 	endpoint          string
 	overrideInCluster bool
 	verbose           bool
+	gitHostTypes      map[string]string
+}
+
+// Only want AutomaticEnv to be called once!
+func init() {
+	// Setup flag to env mapping:
+	//   config-repo => GITOPS_CONFIG_REPO
```
Comment (bigkevmcd) [inline cmd/gitops/root/cmd.go:44]: Thanks for adding this!

**Review by bigkevmcd — APPROVED** [review]
Looks good!

**Comment by efenex** [comment]
We have a setup with an onprem GitLab server running SSH on a non-standard port (2222 as opposed to 22) and I believe it is currently not possible to get this to work with the code in this PR.

I try the installation with something like:

```
gitops install --config-repo=git@gitlab.acme.org:2222/deployments/gitops-weaveworks.git --namespace gitops --git-host-types="gitlab.acme.org:2222=gitlab"
```

And it fails with:

```
Error: error creating git clients: error normalizing url: could not get provider name from URL git@gitlab.acme.org:2222/deployments/gitops-weaveworks.git: no git providers found for "git@gitlab.acme.org:2222/deployments/gitops-weaveworks.git"
```

Would be nice if there was a way to specify how SSH authentication should take place, as I'm not sure how to specify an SSH key or a PAT for this URL at all.

**Comment by foot** [comment]
Hi @efenex ! Thanks for testing this out.

Hmmm, just checking you've tried `--git-host-types="gitlab.acme.org=gitlab"` without the port?

Custom ports is indeed something we overlooked in this PR 🤦 .

**Comment by foot** [comment]
This code is a bit suspect https://github.com/weaveworks/weave-gitops/blob/main/pkg/gitproviders/repo_url.go#L131 and should probably be replaced with something more robust like https://pkg.go.dev/github.com/whilp/git-urls

To workaround, things to try:

Use the full ssh:// form:
```
gitops install --config-repo=ssh://git@gitlab.acme.org:2222/deployments/gitops-weaveworks.git --namespace gitops --git-host-types="gitlab.acme.org:2222=gitlab"
```

**Pull request merged**